src/swap.rs

```diff
@@ -0,0 +1,140 @@
+                amount_rgb,
+                amount_msats,
+            },
+        }
+    }
+
+    pub fn is_buy(&self) -> bool {
+        matches!(self, SwapType::BuyAsset { .. })
+    }
+
+    pub fn amount_msats(&self) -> u64 {
+        match self {
+            SwapType::BuyAsset { amount_msats, .. } | SwapType::SellAsset { amount_msats, .. } => {
+                *amount_msats
+            }
+        }
+    }
+    pub fn amount_rgb(&self) -> u64 {
+        match self {
+            SwapType::BuyAsset { amount_rgb, .. } | SwapType::SellAsset { amount_rgb, .. } => {
+                *amount_rgb
+            }
+        }
+    }
+
+    pub fn side(&self) -> &'static str {
+        match self {
+            SwapType::BuyAsset { .. } => "buy",
+            SwapType::SellAsset { .. } => "sell",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SwapString {
+    pub asset_id: ContractId,
+    pub swap_type: SwapType,
+    pub expiry: u64,
+    pub payment_hash: PaymentHash,
+}
+
+impl std::str::FromStr for SwapString {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.split("/");
+        let amount = iter.next();
+        let asset_id = iter.next();
+        let side = iter.next();
+        let price = iter.next();
+        let expiry = iter.next();
+        let payment_hash = iter.next();
+
+        if payment_hash.is_none() || iter.next().is_some() {
+            return Err("Parsing swap string: wrong number of parts");
+        }
+
+        let amount = amount.unwrap().parse::<u64>();
+        let asset_id = ContractId::from_str(asset_id.unwrap());
+        let price = price.unwrap().parse::<u64>();
+        let expiry = expiry.unwrap().parse::<u64>();
+        let payment_hash = hex_str_to_vec(payment_hash.unwrap())
+            .and_then(|vec| vec.try_into().ok())
+            .map(|slice| PaymentHash(slice));
+
+        if amount.is_err()
+            || asset_id.is_err()
+            || price.is_err()
+            || expiry.is_err()
+            || payment_hash.is_none()
+        {
+            return Err("Parsing swap string: unable to parse parts");
+        }
+
+        let amount = amount.unwrap();
+        let asset_id = asset_id.unwrap();
+        let price = price.unwrap();
+        let expiry = expiry.unwrap();
+        let payment_hash = payment_hash.unwrap();
+
+        if amount == 0 || price == 0 || expiry == 0 {
+            return Err("Parsing swap string: amount, price and expiry should be non-zero");
+        }
+
+        let swap_type = match side {
+            Some("buy") => SwapType::BuyAsset {
+                amount_rgb: amount,
+                amount_msats: amount * price,
+            },
+            Some("sell") => SwapType::SellAsset {
+                amount_rgb: amount,
+                amount_msats: amount * price,
+            },
+            _ => {
+                return Err("Invalid swap type");
+            }
+        };
```

Comment:
```suggestion
        let amount_msats = amount * price; 
        let swap_type = match side {
            Some("buy") => SwapType::BuyAsset {
                amount_rgb: amount,
                amount_msats,
            },
            Some("sell") => SwapType::SellAsset {
                amount_rgb: amount,
                amount_msats,
            },
            _ => {
                return Err("invalid swap type");
            }
        };
```

of course please adapt the suggestion with the suggested renaming. also, could you please implement the `FromStr` trait for the enum instead of matching the string here?

Reply:
I don't think I can implement `FromStr`, because from "buy" or "sell" I don't have the info about `amt_msat` and `asset_amount`. I could create `SwapType` with 0 `amt_msat` and `asset_amount`, imho it's not very clean, but depends what you prefer. What do you think?

Reply:
You're right, the `FromStr` is not doable, it was an oversight, sorry. Please ignore that request